### ISSUE TITLE
Support Nonce in dPoP request header in token endpoint

### ISSUE THREAD

**Issue by daniel-idrees** [eu-digital-identity-wallet/eudi-lib-jvm-openid4vci-kt#330]
**Reference:** https://datatracker.ietf.org/doc/html/rfc9449

In `TokenEndpointClient.kt`,  `fun requestAccessToken(..)`: 

```
            httpClient.submitForm(tokenEndpoint.toString(), formParameters) {
                dPoPJwtFactory?.let { factory ->
                    dpop(factory, tokenEndpoint, Htm.POST, accessToken = null, nonce = null)
                }
```
The nonce is set to null. 

Can we allow `fun AuthorizationRequestPrepared.authorizeWithAuthorizationCode()` in `AuthorizeIssuance.kt` to accept a nonce as nullable string parameter and then allow `fun requestAccessToken()` to take the same parameter to be set in the dPoP request header?

Can I raise a PR with the above changes that you can review?

**Comment by babisRoutis** [comment]
Dear @daniel-idrees 

For starters, thank you for raising this and certainly community PRs a are welcomed. 

Indeed, currently, library doesn't support neither, 
* [Authorization Server Provided DPoP Nonce](https://datatracker.ietf.org/doc/html/rfc9449#name-authorization-server-provid), nor
* [Resource Server Provided DPoP Nonce](https://datatracker.ietf.org/doc/html/rfc9449#name-resource-server-provided-no), nor
* [Authorization Code Binding to a DPoP Key](https://datatracker.ietf.org/doc/html/rfc9449#name-authorization-code-binding-)

I assume that you refer to the first case, so lets focus on **Authorization Server Provided DPoP Nonce** .
To support this, I think that library should be able to:

1. Use a Authorization Server-provided DPoP Nonce &
2. Update this DPoP Nonce, in case Authorization Server provides a fresh value
3. Possibly, retry a call with a new DPoP Nonce


I would try to provide some more details on those three tasks.

**Comment by babisRoutis** [comment]
I starting point would be the `TokenEndpointClient`

All public methods should accept an optional `dpopNonce` and return `Result<Pair<TokenResponse, Nonce?>>`

The right part of the pair, is either what was provided in the input (if any) or a fresh dPoPNonce value provided by the authorization server.


An implementation of `requestAccessToken()` function, that would support recovery due to error `use_dpop_nonce`
could be:

```kotlin
private suspend fun requestAccessToken(
        params: Map<String, String>,
        dpopNonce: Nonce?
    ): Pair<TokenResponse, Nonce?> {

        suspend fun requestInternal(existingDpopNonce: Nonce?): Pair<TokenResponseTO, Nonce?> {

            val response = ktorHttpClientFactory().use { httpClient ->
                val formParameters = Parameters.build {
                    params.entries.forEach { (k, v) -> append(k, v) }
                }
                httpClient.submitForm(tokenEndpoint.toString(), formParameters) {
                    dPoPJwtFactory?.let { factory ->
                        dpop(factory, tokenEndpoint, Htm.POST, accessToken = null, nonce = existingDpopNonce?.value)
                    }
                    generateClientAttestationIfNeeded()?.let(::clientAttestationHeaders)
                }
            }

            return when {
                response.status.isSuccess() -> {
                    val responseTO = response.body<TokenResponseTO.Success>()
                    val newDopNonce = response.headers["DPoP-Nonce"]?.let { nonce -> Nonce(nonce) }
                    responseTO to (newDopNonce ?: existingDpopNonce)
                }

                response.status == HttpStatusCode.BadRequest -> {
                    val errorTO = response.body<TokenResponseTO.Failure>()
                    val newDopNonce = response.headers["DPoP-Nonce"]?.let { nonce -> Nonce(nonce) }
                    when {
                        errorTO.error == "use_dpop_nonce" && newDopNonce != null -> requestInternal(newDopNonce)
                        else -> errorTO to (newDopNonce ?: existingDpopNonce)
                    }
                }

                else -> throw AccessTokenRequestFailed("Token request failed with ${response.status}", "N/A")
            }
        }

        val (responseTO, newDopNonce) = requestInternal(dpopNonce)
        return responseTO.tokensOrFail(clock) to newDopNonce
    }
```

**Comment by babisRoutis** [comment]
Dear @daniel-idrees 

In general, I think it won't be trivial to add support for authorization server provided dpopNonce values.

Changes required:
- Change token endpoint client as described above
- Change PAR endpoint client, in a similar way. 
- Change `AuthorizedRequest` to add a new optional `dpopNonce` attribute
- Change `RefreshToken` to return a new `AuthorizedRequest` updated with new `access_token`  and in addition a new `dpopNonce`
- Change `AuthorizationRequestPrepared`to add a new optional `dpopNonce` attribute. It would be populated if PAR + dpop is being used

For that reason I 've created an initial feature branch [here](https://github.com/niscy-eudiw/eudi-lib-jvm-openid4vci-kt/tree/feat/dpopNonce), but keep in mind that this is a fork of the official repo. 

To my understanding there is no need to add a a new parameter to `AuthorizationRequestPrepared.authorizeWithAuthorizationCode()`. Can you elaborate on this?

Finally, If you want a create a PR, you can use the aforementioned branch as a basis.
Otherwise if this change seems too complex, please let me know to create a PR (using the above branch)

**Comment by daniel-idrees** [comment]
Hi @babisRoutis , 

Thank you for the response. Can we use our own forked repository to create a PR? 

Also, we tried this implementation which you check here: https://github.com/german-first-iteration/eudi-lib-jvm-openid4vci-kt/compare/main...german-first-iteration:eudi-lib-jvm-openid4vci-kt:adding-dpnounce-to-credential-endpoint-request

For the time being, the idea is to accept dpop nonce by the public methods of the library so that we can add it in the header for the token endpoint call and credential endpoint call. 

The implementation that you shared above also meets the criteria as per https://datatracker.ietf.org/doc/html/rfc9449 and we can also work on this and create a PR to re-request the the access token when the error "use_dpop_nonce" is received, with the nonce in the response header. 

What do you think? 

Thanks

**Comment by babisRoutis** [comment]
Dear @daniel-idrees 

> Thank you for the response. Can we use our own forked repository to create a PR?
> 

Yes. That's the idea. Fork the repo, in the forked repo don't change the main (to be able to sync with origin), create a branch and place a PR to origin repo, usually targeting the main branch.


> Also, we tried this implementation (working fine) which you can check here: [german-first-iteration/eudi-lib-jvm-openid4vci-kt@main...german-first-iteration:eudi-lib-jvm-openid4vci-kt:adding-dpnounce-to-credential-endpoint-request](https://github.com/german-first-iteration/eudi-lib-jvm-openid4vci-kt/compare/main...german-first-iteration:eudi-lib-jvm-openid4vci-kt:adding-dpnounce-to-credential-endpoint-request)
> 
> For the time being, the idea is to accept dpop nonce by the public methods of the library so that we can add it in the header for the token endpoint call and credential endpoint call.
> 
> The implementation that you shared above also meets the criteria as per https://datatracker.ietf.org/doc/html/rfc9449 and we can also work on this and create a PR to re-request the the access token when the error "use_dpop_nonce" is received, with the nonce in the response header. (We can do this separately in a different PR)
> 

I think that the branch that I shared has already implemented the retry on `use_dpop_nonce` both for :

- [Token endpoint](https://github.com/niscy-eudiw/eudi-lib-jvm-openid4vci-kt/blob/164a49a266b234c303688af6808d8c6a519aaca6/src/main/kotlin/eu/europa/ec/eudi/openid4vci/internal/http/TokenEndpointClient.kt#L231) and
- [PAR endpoint](https://github.com/niscy-eudiw/eudi-lib-jvm-openid4vci-kt/blob/164a49a266b234c303688af6808d8c6a519aaca6/src/main/kotlin/eu/europa/ec/eudi/openid4vci/internal/http/AuthorizationEndpointClient.kt#L276),  (BTW, was missing DPoP support)

In addition, on those two cases we do update the `dpopNonce`, also in the case of `200` .
No matter the case, the single `dpopNonce` on the client side is kept within the `AuthorizedRequest` (which is updated with each interaction, since the implementation is stateless).

To my understanding, we cover all the cases where `dpopNonce` is updated by the authorization server either as a result of a `use_dpop_nonce` error or when it is just being updated with a `200` response. 

If for whatever reason, you want to manually update the dpopNonce, then you can just update the `AuthorizedRequest`

Bottom line, I propose the following:
- Let me create a PR out of the branch that I shared
- It would be really helpful, if you can check it against your authorization server and provide your feedback/comments

**Comment by babisRoutis** [comment]
@daniel-idrees Created the PR. 
If you will, please provide your comments there

**Comment by dzarras** [comment]
Dear @daniel-idrees ,

We have created a PR to address this issue. Could you please check #332 and let us know of any comments?

Kind regards.

**Comment by daniel-idrees** [comment]
Hi @dzarras 

I able to test https://github.com/eu-digital-identity-wallet/eudi-lib-jvm-openid4vci-kt/pull/332 and for me it is working. 

Having said that, we also need dpop nonce which is received in the header of the tokenendpoint response, to be passed to the header request of credential endpoint. We tried these changes which worked for us https://github.com/german-first-iteration/eudi-lib-jvm-openid4vci-kt/pull/2/files that you can have a look. Should I raise separate issue for this? I can also create a PR with these changes too. 

Secondly, will you also add the dpopnonce support to the iOS version of the open4vci repository? 

cc: @babisRoutis

**Comment by dzarras** [comment]
> Hi @dzarras
> 
> I able to test #332 and for me it is working.
> 
> Having said that, we also need dpop nonce which is received in the header of the tokenendpoint response, to be passed to the header request of credential endpoint. We tried these changes which worked for us https://github.com/german-first-iteration/eudi-lib-jvm-openid4vci-kt/pull/2/files that you can have a look. Should I raise separate issue for this? I can also create a PR with these changes too.
> 
> Secondly, will you also add the dpopnonce support to the iOS version of the open4vci repository?
> 
> cc: @babisRoutis

Hi @daniel-idrees, Thanks for the input. 

The current PR addresses [Authorization Server-Provided Nonce](https://datatracker.ietf.org/doc/html/rfc9449#name-authorization-server-provid) only. I'd prefer if we'd track [Resource Server-Provided Nonce](https://datatracker.ietf.org/doc/html/rfc9449#name-resource-server-provided-no) support in a new issue and implement it in a separate PR.

By the way, are you targeting an an Issuer that is acting as an Authorization Server at the same time? Is this why you want to be able to pass the DPoP Nonce provided by the Token Endpoint of the Authorization Server to the Credential Endpoint of the Issuer?

We try to have feature parity between the Kotlin and Swift libraries. We will try to add support in the Swift library at a later date.

Kind regards.

cc: @babisRoutis

**Comment by babisRoutis** [comment]
Hi @daniel-idrees  & @dzarras 

I would agree that supporting Resource Server-provided Nonce should be addressed in a separate issue/PR.

Interactions (wallet to Credential issuer) with the following protocol endpoints would be in scope:

- Credential endpoint
- Deferred endpoint
- Notification endpoint
- `c_nonce` endpoint (introduced with latest OpenId4VCI)

Implementation could be similar to this one. 
1. Add a new optional attribute,named `credentialIssuerDPoPNonce`, to `AuthorizedRequest` 
2. Clients of the aforementioned endpoints should be able to recover a `401` status response

```text
HTTP/1.1 401 Unauthorized
 WWW-Authenticate: DPoP error="use_dpop_nonce", \
   error_description="Resource server requires nonce in DPoP proof"
 DPoP-Nonce: eyJ7S_zG.eyJH0-Z.HX4w-7v
```
3. Clients of the aforementioned endpoints  should be able to extract a fresh `credentialIssuerDPoPNonce` from a status 200 response.

@dzarras 
1. Please merge the #332 
2. Open a new issue describing Resource Server-provided DPoP Nonce

If it is critical for @daniel-idrees proceed with a new PR as described above.


PS: I think that resource server provided DPoP nonce is a corner case. Implementers would have to catch up with OpenId4VCI especially after the removal of `c_nonce` from the token endpoint, which may be followed also by the removal of `c_nonce` from credential response and the use of the newly defined nonce endpoint.
